internal/controller/promotionstrategy_controller.go

```diff
@@ -341,14 +342,64 @@ func (r *PromotionStrategyReconciler) calculateStatus(ps *promoterv1alpha1.Promo
 // (Note.DrySha if set, otherwise Proposed.Dry.Sha). If they differ, the CTPs with
 // different values need to reconcile to fetch updated git notes or proposed dry sha. This is needed
 // because GitHub doesn't send webhooks when git notes are pushed.
-// Returns true if any CTPs were enqueued for reconciliation.
-func (r *PromotionStrategyReconciler) enqueueOutOfSyncCTPs(ctx context.Context, ctps []*promoterv1alpha1.ChangeTransferPolicy) bool {
+// Rate limiting: Only enqueues a CTP once per 15 seconds. If rate limited, schedules a delayed enqueue.
+func (r *PromotionStrategyReconciler) enqueueOutOfSyncCTPs(ctx context.Context, ctps []*promoterv1alpha1.ChangeTransferPolicy) {
 	logger := log.FromContext(ctx)
 
 	if len(ctps) == 0 {
-		return false
+		return
 	}
 
+	// Initialize state map lazily
+	if r.enqueueStates == nil {
+		r.enqueueStateMutex.Lock()
+		if r.enqueueStates == nil {
+			r.enqueueStates = make(map[client.ObjectKey]*ctpEnqueueState)
+		}
+		r.enqueueStateMutex.Unlock()
+	}
+
+	// Lazy initialize map and start background cleanup timer
+	if r.enqueueStates == nil {
+		r.enqueueStates = make(map[client.ObjectKey]*ctpEnqueueState)
+
+		// Start self-rescheduling cleanup timer to prevent memory leak from deleted CTPs.
+		//
+		// Memory footprint per entry (64-bit system, measured with unsafe.Sizeof):
+		//   - client.ObjectKey (2 strings): ~96 bytes
+		//       * Struct: 32 bytes (2 string headers, 16 bytes each)
+		//       * String content: namespace (32 chars) + name (32 chars) = 64 bytes
+		//   - *ctpEnqueueState pointer: 8 bytes
+		//   - ctpEnqueueState struct: 32 bytes
+		//       * time.Time: 24 bytes
+		//       * bool: 1 byte + 7 bytes padding = 8 bytes
+		//   - Map overhead: ~8 bytes per entry
+		//   Total: ~144 bytes per CTP
+		//
+		// Memory bounds (assuming 32-char namespace and name):
+		//   - 100 stale entries = ~14 KB
+		//   - 1,000 stale entries = ~144 KB
+		//   - 10,000 stale entries = ~1.4 MB
+		//
+		// With 1 hour cleanup interval, worst case is 1 hour of deleted CTPs in memory.
```

Comment:
Probably okay to move the memory analysis to the PR description.

Reply:
ehh.. sure I kind of liked it there but can just remove the whole thing if you don't like it, it nice to know what is in each cache that we have. I imagine something like that would be really nice in where argocd had an in memory cache